[lldb-dap] Add multi-session support with shared debugger instances

qxy11 opened the pull request:
## Summary:
This change introduces a DAPSessionManager to enable multiple DAP sessions to share debugger instances when needed, for things like child process debugging and some scripting hooks that create dynamically new targets.

Changes include:
- Add DAPSessionManager singleton to track and coordinate all active DAP sessions
- Support attaching to an existing target via its globally unique target ID (targetId parameter)
- Share debugger instances across sessions when new targets are created dynamically
- Refactor event thread management to allow sharing event threads between sessions
- Add eBroadcastBitNewTargetCreated event to notify when new targets are created
- Extract session names from target creation events
- Defer debugger initialization from 'initialize' request to 'launch'/'attach' requests. The only time the debugger is used currently in between its creation in `InitializeRequestHandler` and  the `Launch` or `Attach` requests is during the `TelemetryDispatcher` destruction call at the end of the `DAP::HandleObject` call, so this is safe.

This enables scenarios when new targets are created dynamically so that the debug adapter can automatically start a new debug session for the spawned target while sharing the debugger instance.

## Tests:
The refactoring maintains backward compatibility. All existing DAP test cases pass.

Also added a few basic unit tests for DAPSessionManager
```
>> ninja DAPTests
>> ./tools/lldb/unittests/DAP/DAPTests
>>./bin/llvm-lit -v ../llvm-project/lldb/test/API/tools/lldb-dap/
```

qxy11 committed:
[lldb-dap] Add multi-session support with shared debugger instances

Summary:
This change introduces a DAPSessionManager to enable multiple concurrent
lldb-dap sessions and allow them to share debugger instances when needed.

Key changes:
- Add DAPSessionManager singleton to track and coordinate all active DAP sessions
- Support attaching to an existing target via unique target ID (targetId parameter)
- Share debugger instances across sessions when spawning new targets (e.g., GPU debugging)
- Refactor event thread management to allow sharing event threads between sessions
- Add eBroadcastBitNewTargetCreated event to notify when new targets are spawned
- Extract session names from target creation events for better identification
- Defer debugger initialization from 'initialize' request to 'launch'/'attach' requests

This enables scenarios where a native process spawns a new target (like a child process)
and the debug adapter can automatically start a new debug session for the spawned
target while sharing the parent's debugger instance.

Tests:
The refactoring maintains backward compatibility. All existing test cases pass.

walter-erquinigo commented on lldb/include/lldb/API/SBTarget.h:
```diff
@@ -69,6 +70,8 @@ class LLDB_API SBTarget {
   static lldb::SBModule GetModuleAtIndexFromEvent(const uint32_t idx,
                                                   const lldb::SBEvent &event);
 
+  static const char *GetSessionNameFromEvent(const SBEvent &event);
```
add documentation mentioning which kind of events carry a session name

walter-erquinigo commented on lldb/tools/lldb-dap/DAPSessionManager.cpp:
```diff
@@ -0,0 +1,168 @@
+
+DAPSessionManager &DAPSessionManager::GetInstance() {
+  // NOTE: intentional leak to avoid issues with C++ destructor chain
+  // Use std::call_once for thread-safe initialization.
```
just remove this comment

walter-erquinigo commented on lldb/tools/lldb-dap/DAPSessionManager.h:
```diff
@@ -0,0 +1,119 @@
+  std::thread m_event_thread;
+};
+
+/// Global DAP session manager.
```
State here what's the purpose of this object and the most relevant aspects of it that people should know about

qxy11 commented on lldb/tools/lldb-dap/DAPSessionManager.h:
```diff
@@ -0,0 +1,119 @@
+
+  /// Optional map from target ID to shared debugger set when the native
+  /// process creates a new target.
+  std::map<uint32_t, lldb::SBDebugger> m_target_to_debugger_map;
```
I'm a bit confused what the thread id here is, do you mean `target id -> SBTarget`?

qxy11 commented on lldb/tools/lldb-dap/DAPSessionManager.h:
```diff
@@ -0,0 +1,119 @@
+
+  /// Wait for all sessions to finish disconnecting.
+  /// Returns an error if any client disconnection failed, otherwise success.
+  llvm::Error WaitForAllSessionsToDisconnect();
```
`DisconnectAllSessions()` should trigger the disconnection, while `WaitForAllSessionsToDisconnect()` blocks and waits for the sessions to disconnect and unregister.

walter-erquinigo replied:
mention that in the documentation :)

qxy11 committed:
Add GetSessionName, remove GetSessionNameFromEvent

qxy11 committed:
Use StoreTargetById/RetrieveTargetById and add more comments

qxy11 committed:
darker lint

JDevlieghere commented:
I haven’t had a chance to take a look at this yet, but given that this is a significant change, please give me (and other contributors) a bit more time to review this before merging this.

ashgti commented on lldb/tools/lldb-dap/DAPSessionManager.h:
```diff
@@ -0,0 +1,125 @@
+  llvm::Error WaitForAllSessionsToDisconnect();
+
+  /// Store a target for later retrieval by another session.
+  void StoreTargetById(uint32_t target_id, lldb::SBTarget target);
```
I'm not sure I understand this part, why do we need to store them separately and the `TakeTargetById`, why does that remove them from the list of targets?

walter-erquinigo replied:
Now that I think of it, you don't need to this mapping.

Assuming that a DAP debug session has a single debugger instance, which I think is true 99.999% of the cases, whenever you want to attach to a child target by ID, you can get the Debugger instance of any existing target, then you can ask its debugger for its target list, and then you you do a linear scan on it finding the ID you want. 
Then you can avoid doing all this tracking.

clayborg replied:
There is a function in the `lldb::SBDebugger` that can already find you a target by and a debugger by ID (and debugger IDs start with 1 and increment, so it would be easy to iterate if needed:
```
  static lldb::SBDebugger lldb::SBDebugger::FindDebuggerWithID(int id);
  lldb::SBTarget lldb::SBDebugger::FindTargetByGloballyUniqueID(lldb::user_id_t id);
```
So you should be able to find what you need without tracking this in `StoreTargetById` as Walter suggested.

You can also get the debugger ID easily if needed:

```
(lldb) script lldb.debugger.GetID()
1
```
And this API is availabe from C++.

ashgti commented on lldb/tools/lldb-dap/DAPSessionManager.h:
```diff
@@ -0,0 +1,125 @@
+  llvm::Error WaitForAllSessionsToDisconnect();
+
+  /// Store a target for later retrieval by another session.
+  void StoreTargetById(uint32_t target_id, lldb::SBTarget target);
```
Do we need to expose this in some way to the scripting interface? We expose a few custom lldb commands for triggering custom DAP events or triggering a `startDebugging` reverse request. Being able to store a target and trigger a new `startDebugging` reverse request seems like one of the ways this would be useful.

ashgti commented on lldb/tools/lldb-dap/DAPSessionManager.cpp:
```diff
@@ -0,0 +1,169 @@
+                                        lldb::SBTarget target) {
+  std::lock_guard<std::mutex> lock(m_sessions_mutex);
+  m_target_map[target_id] = target;
+}
```
Do we need to make any assumptions about the target? E.g. should we assert that its paused? Is it valid for the DAP to continue using the SBTarget? If not, we may want this to take a `DAP &dap` param and remove the target from the DAP instance.

ashgti commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -1390,104 +1481,151 @@ void DAP::EventThread() {
             }
           }
+        } else if (event_mask & lldb::SBTarget::eBroadcastBitNewTargetCreated) {
+          auto target = lldb::SBTarget::GetTargetFromEvent(event);
```
What if you have a `launchCommand`/`attachCommand` that creates a target?

Should we ignore this event while performing the `launch`/`attach` commands?

ashgti reviewed (commented):
Can we also add a test in test/API/tools/lldb-dap for this? Maybe in the test/API/tools/lldb-dap/startDebugging folder?

qxy11 committed:
Use lldb::user_id_t for target_id

Summary:
- Changed targetId type from uint32_t to lldb::user_id_t
- Added comprehensive documentation for StoreTargetById/TakeTargetById

qxy11 committed:
Remove StoreTarget/TakeTarget

walter-erquinigo commented on lldb/include/lldb/Target/Target.h:
```diff
@@ -622,6 +623,30 @@ class Target : public std::enable_shared_from_this<Target>,
+  ///
+  /// \return
+  ///     The target session name for this target.
+  const std::string &GetTargetSessionName() { return m_target_session_name; }
```
```suggestion
  llvm::StringRef GetTargetSessionName() { return m_target_session_name; }
```

You should use StringRef. It can represent different kinds of strings automatically.

walter-erquinigo commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -1289,13 +1293,108 @@ protocol::Capabilities DAP::GetCustomCapabilities() {
+    if (!debugger.IsValid()) {
+      return llvm::createStringError(
+          "Unable to find existing debugger for debugger ID");
+    }
```
this doesn't look very clean.
What about creating two `InitializeDebugger` functions? One doesn't take parameters and the other one takes forcefully those two ids?

walter-erquinigo commented on lldb/tools/lldb-dap/EventHelper.cpp:
```diff
@@ -295,4 +307,267 @@ void SendMemoryEvent(DAP &dap, lldb::SBValue variable) {
   dap.Send(protocol::Event{"memory", std::move(body)});
 }
 
+// Note: EventThread() is architecturally different from the other functions in
```
thanks for this comment

ashgti commented on lldb/tools/lldb-dap/EventHelper.cpp:
```diff
@@ -295,4 +307,267 @@ void SendMemoryEvent(DAP &dap, lldb::SBValue variable) {
+              dap_instance = active_instances[0];
+          }
+
+          if (dap_instance) {
```
Should we report an error if we made it here without a dap_instance? That seems like an internal logic issue, right?

clayborg reviewed (approved):
Looks like feedback was followed up on, lets start with this and iterate on it!

qxy11 committed:
Add basic startDebugging test

qxy11 committed:
Moving Event Thread out of DAP

qxy11 committed:
Target::GetTargetSessionName returns llvm::StringRef

qxy11 committed:
Two InitializeDebugger functions

qxy11 committed:
Simplify attach json args

qxy11 committed:
Send reverse request from parent DAP instance and report errors

qxy11 committed:
Format

clayborg commented:
@Michael137 any ideas on how to track down what is going wrong here? How do we run these tests?

qxy11 commented:
https://github.com/llvm/llvm-project/pull/169744 should be able to fix the cross-project-test errors
